packages/emnapi/src/init.ts

```diff
@@ -179,7 +179,7 @@ mergeInto(LibraryManager.library, {
       return emnapiExports
     }
 
-    addOnInit(function (Module) {
+    __ATINIT__.push(function (Module) {
```

Comment:
Wouldn't it be easier to do all this init from C/C++ rather than by dealing with JS helpers?

For example, this callback could be just another `__attribute__((constructor))` instead of adding it as a separate init from JS, and then all those surrounding calls would probably be simpler & smaller when express via napi.h too.

Reply:
This callback will not be called until [`addRunDependency("emnapi-init")`](https://github.com/toyobayashi/emnapi/blob/main/packages/emnapi/src/init.ts#L118) is removed, it's asynchronous

Reply:
Hm why? Is there anything in actual N-API that is asynchronous?

Reply:
https://github.com/toyobayashi/emnapi/blob/b7501570eebd9acfbe362aeadd7168e7df688e96/packages/emnapi/src/init.ts#L117-L129

https://github.com/toyobayashi/emnapi/blob/b7501570eebd9acfbe362aeadd7168e7df688e96/packages/emnapi/src/init.ts#L26

`emnapiRuntimeInit` (async)  then `emnapiInit` (`__ATINIT__.push()`)

Reply:
`emnapiInit` depends `emnapiRt` and `emnapiCtx`, which is asynchronously initialized in `emnapiRuntimeInit`

Reply:
Hm, but then, C/C++, including `__attribute__((constructor))`, also won't be called until `addRunDependency` is removed, so the effect should be the same?

Reply:
But we always need to create `Env` and export `emnapiExport` on JavaScript side, current implementation can easily use closure to save variables. (I found out Node.js official `node_api.h` uses `__attribute__((constructor))` only on native platforms, but export `napi_register_wasm_v1` on wasm, during initialization the host environment should pass `env` and `export` to it.)

https://github.com/nodejs/node/blob/513c15147cb44d2d27073e520820146185c45638/src/node_api.h#L76-L107